### ISSUE TITLE
Upgrade to bootstrap 4

### ISSUE THREAD

**Issue by samdark** [yiisoft/yii2-bootstrap#77]

**Comment by df2** [comment]
How about creating a seperate repo `yii2-bootstrap4`, which developers can specify in `composer.json`?

**Comment by cebe** [comment]
not sure this is needed, we have bootstrap extension 2.0.x branch for bootstrap 3 and go with bootstrap 4 in 2.1.

**Comment by freezy-sk** [comment]
@cebe I think that upgrade to Bootstrap 4 should bump package version to 3.0.0

**Comment by cebe** [comment]
would be good if you both could cover your suggestions with arguments :)

**Comment by freezy-sk** [comment]
@cebe My suggestion is based on simple semantic versioning.
Bootstrap 3 will be still alive, so if they change something in that branch, 2.x.y is here to reflect all that changes.
Bootstrap 4 is refactored with so many breaking changes that 3.0.0 of this package should be nice starting point for maintaining all new features which will be introduced during B4 lifecycle.

**Comment by cebe** [comment]
Yii does not follow semantic versioning, so 2.1 can break BC. For patch releases we do not distinguish between bugs and features.  See https://github.com/yiisoft/yii2/blob/master/docs/internals/versions.md#yii-versioning

**Comment by freezy-sk** [comment]
@cebe Yes, I know how is versioned main framework repository, but I still think that it's much more clearer if will be package incremented to 3.0.0. instead of 2.1.0 because it's like whole new package. It will be also easier to maintain 2.x.y branch with new features and bugfixes for B3 (and users which can't easy upgrade frontend of their projects) and 3.x.y will be for B4.

**Comment by df2** [comment]
It's just that Yii 2 has been all about seamless integration with Bootstrap. There's not enough added value (flexbox may be useful?) to update all our components to the new ones. In addition, our theme is heavily customized with Less, and these would all have to be re-written. Yii 2.1 is still Yii 2, and should be a few minute refactor to upgrade, not an entire re-write of our views. When Bs4 is actually launched, users of 2.0 or 2.1 can have the option of specifying it in `composer.json`, and use the new widgets.

Alternatively, rename this repo to `yii2-bootstrap3` and have that override `yii2-bootstrap` (which is now Bs4).

**Comment by Sammaye** [comment]
LESS and SASS are very similar you won't fnd much rewriting to be done.

BS4 will be really good, especially in better dealing with responsive typography after the IE8 drop, finally we will actually be able to natively have real responsive typography in BS instead of using breakpoints on certain headers etc.

Looking at first glance the components don't look like they have been massively rewritten on the frontend, they still use the same classes in most places. Most of BS4 is on the backend. They keep saying about this grid tier for better mobile handling (which would be awesome, I did this manually) but all I see is the BS3 grid tiers so I don't think the docs are even truly documenting what they have done yet.

**Comment by str2code** [comment]
唔，相当期待`bootstrap4`

**Comment by pasci84** [comment]
I wouldn't bind Yii to a fixed version of bootstrap as others (eg Joomla) do. As @df2 stated it would be very clean if there were two repos each extending Yii with a bootstrap-version. That way they could be maintained independently and the 3-version dropped at some point. Also the planning of the yii-repos allows this, as bootstrap is not in the core but in its own repo. Yes, there would be some duplication of code as the bootstrap versions are semantically close, but the gain for devs and admins would be huge...

Conclusion: I totally agree with the idea to specify this in the json-file. Default should be bs4 and optionally we could change it back to bs3 and be sure that everything runs fine the way it was before.

**Comment by Sammaye** [comment]
+1 ideally we rename this repo to yii2-bottstrap3 and make a new yii2-bootstrap4 repo

**Comment by realtebo** [comment]
Please, duplicated yii2-bootstrap as -3 and -4, and, please again, release an 'alpha' version ASAP. 

Move a COMPLEX webapp from 2.0.x to 2.1 AND from BS3 to BS4 is NOT a easy task in a lot of real world applications, with a lot of custom widgets and extensions. We need to do extensive alpha-work and testing on each our apps

**Comment by lukos** [comment]
Like others have said, many existing projects are using yii-bootstrap version "*". If you upgrade the project without renaming it, anyone running composer update on an existing project will suddenly cause a lot of breakage and might not know why - I think lots of questions will appear on the forums.

I can't see any reason not to create a yii-bootstrap4 project (easier to understand than yii-bootstrap3!) and keep them both separate.

**Comment by YasserHassan** [comment]
Bootstrap version in a project is determined by other components besides the framework such as the template and jQuery extensions used. Developers should be able to decide the time to move from BS 3 to 4 in the time they see fit. They might be held back from upgrading because one or more of those components have not yet adopted BS4. That's why being able to decide which BS version to use with the framework and changing that at any time (even to downgrade from v4 to v3 later for any reason) would be very important. Specifying BS version in `composer.json` is probably the best way to go.

**Comment by jacmoe** [comment]
I can't see any reason why you shouldn't upgrade to Bootstrap 4 because the code is much cleaner and it is easier to customize because they reorganized and unnested the css structure quite a bit.
Moving from Less to Sass might be daunting, but you will be powerful.
I expect rapid adoption all over the board.

However, Yii-2-bootstrap should update to Bootstrap 3.5 as soon as possible as they fixed some serious bugs since 3.3..

**Comment by YasserHassan** [comment]
@jacmoe there would be no reason to upgrade to Bootstrap 4 in a stable system if it does not provide an added value. The upgrade will require work to implement, test, and adjust. And before that you need to make sure that each and every component you use (e.g. Select2) has upgraded to BS4 and it's Yii2 package is upgraded to the component's new version too. It could be better for some businesses to spend that time and effort in adding new features to their system and they should be able to select the upgrade time best suiting their situation.

**Comment by jacmoe** [comment]
If that stable system is going to be maintained for the foreseeable future then I think moving from Less to Sass would help tremendously in keeping the code modular and manageable, and - from what I've seen - Bootstrap 4 doesn't change that much interface-wise. It just becomes leaner and much easier to extend/override. I think should bring more time to make an effort in adding new features. :)
But let's see. I am stoked and predict rapid adoption!

**Comment by jacmoe** [comment]
So what I am saying is that it definitely has added value. :)

**Comment by jacmoe** [comment]
One last note which - I hope - demonstrates the benefits of upgrading:

http://v4-alpha.getbootstrap.com/layout/grid/#example-usage

Mixins!

Instead of having Yii2-bootstrap use col-md-3 and other specific classes it could now use semantic classes instead.

~~~
.container {
  max-width: 60em;
  @include make-container();
}
.row {
  @include make-row();
}
.content-main {
  @include make-col();

  @media (max-width: 32em) {
    @include make-col-span(6);
  }
  @media (min-width: 32.1em) {
    @include make-col-span(8);
  }
}
.content-secondary {
  @include make-col();

  @media (max-width: 32em) {
    @include make-col-span(6);
  }
  @media (min-width: 32.1em) {
    @include make-col-span(4);
  }
}
~~~
If the layout changes then - instead of performing a mass-edit all across the project - all it takes is a simple change in one file.
And it would also mean that yii2-bootstrap could be used with Zurb Foundation as well, just by changing the grid/layout mixins :)

**Comment by YasserHassan** [comment]
Not everyone needs to do that and if they do, not everyone needs to do it now. My comment was not about arguing the new version's technical benefits but rather about putting the project owners in control and letting them decide what and when and if they perform that upgrade according to to their project circumstances and priorities.

**Comment by craigpotter** [comment]
Would like too see a way to use Bootstrap 4 with Yii2 We got a project planned to go live mid year. Seems a bit pointless developing this with bootstrap 3.x at the moment as we will need to rewrite it all for live to use BS 4. 

I know Bs4 is still in Beta but I think people should be given a chance to use it at their own risks. BS make it clear not to use it for production. If Yii2 also did the same. I don't think their will be an issue.

**Comment by kop** [comment]
+1 to the @craigpotter message.
I'm having a new project too and I believe that it does not have much sense to start with `3.x` branch. Some option to use BS4 would be really cool.

**Comment by dillix** [comment]
I also want to start new project with BS4 support. Is there any news about yii-bs4 branch?

**Comment by marnap89** [comment]
> Is there any news about yii-bs4 branch?

I ask the same question :)

**Comment by samdark** [comment]
No. Noone touched it.

**Comment by alanwillms** [comment]
@samdark people can start working on it as soon as you decide how to handle the version name.

**Comment by samdark** [comment]
I'd handle it via branch i.e.:

1. Create `2.0` branch from current code. Bugfixes for 2.0.x will be there.
2. `master` will hold BS4 version. Tags will be 3.0.x.

**Comment by cebe** [comment]
@alanwillms you can start working on it and send a pull request, we will handle branching then.

**Comment by dillix** [comment]
Great news!

**Comment by mikehaertl** [comment]
To chime in on the discussion: Whatever decision is made, developers should have free choice. Like:

 * Keep using Yii 2.0.x but upgrade to Bs4
 * Upgrade to Yii 2.1.x but keep using Bs3

I'm also tending towards moving things to another repo like `yii2-boostrap4`. This would show that Yii acts very carefully and tries to avoid messing up existing setups by any means. Because I agree to what @lukos said above: developers may have `"yiisoft/yii2-bootstrap": "*"` in their `composer.json` and may upgrade to Bs4 by accident.

It would also make it very clear that it's BS4. Something that is not obvious from "2.1.x".

What exactly is the advantage of keeping everything in one repo?

**Comment by cebe** [comment]
> What exactly is the advantage of keeping everything in one repo?

the bootstrap4 compatible code of this extension is only a slight modification of bootstrap3 compatible code, so it would rather be a new version than a completely new repo.

**Comment by Sammaye** [comment]
Problem to making a new veersion is BS4 gets rid of IE8 support

**Comment by craigpotter** [comment]
I don't see the problem with making it a new version.... If you don't want to use bootstrap 4, don't use "yiisoft/yii2-bootstrap": "*" in composer. 

Lock in the BS3 version. 

I guess this was the whole point of moving this away from the core so people could do this.

**Comment by Sammaye** [comment]
Well, I guess so long as we make one major version BS3 and then all after
BS4 is would work (thinking of updates)

**Comment by justinvoelker** [comment]
I agree with @craigpotter about keeping it in the same repo.  If someone wants to stay on BS3, they should not use "yiisoft/yii2-bootstrap": "*" in composer.

That said, since that is the default in a new install of Yii, is there anything in a _stock_ Yii install that would break if the repo were updated to BS4?  Projects whose views have not been changed from the Gii-generated pages should not require changing after the repo moves to BS4.  This allows all of the projects with the default views in place to not require manual editing to make them work with BS4.

To those of us with heavily customized projects, we will A) recognize the need to lock the composer version to BS3 or B) update our projects to use BS4.

**Comment by mikehaertl** [comment]
>  If you don't want to use bootstrap 4, don't use "yiisoft/yii2-bootstrap": "*" in composer.

Yes - if you know about this issue and how versions here are matched against Bootstrap. Many developers will not be aware of this. So at least it should be made very clear in the upgrade guide.

**Comment by craigpotter** [comment]
I don't think that will be a problem. 

It could be added to the [README.md](https://github.com/yiisoft/yii2-bootstrap/blob/master/README.md), also the [yii2-bootstrap docs](https://github.com/yiisoft/yii2-bootstrap/blob/master/docs/guide/README.md). 

I don't think the guys will have an issue also adding a note to the getting started guide as well - http://www.yiiframework.com/doc-2.0/guide-start-installation.html#installing-via-composer

**Comment by mikehaertl** [comment]
> the bootstrap4 compatible code of this extension is only a slight modification of bootstrap3 compatible  code, so it would rather be a new version than a completely new repo.

Still that's not really a reason why everything should be in one repo. This extension is about Bootstrap and Bs3 and Bs4 are two incompatible frameworks, each with different markup.

What if you want to use both, Bs3 + Bs4 in a single project? For example your frontend should be updated to Bs4, while your backend module should stick with Bs3. 

With a separate repo (and `yii\bootsrap4` namespace) this is easy. All in one repo makes it much harder. I still wonder where exactly is the problem if we have it in a separate repo? 

And the argument "we can document the upgrade path" can also be used in the opposite direction: The well informed developer is never a problem. He will read the upgrade guide and know how to get Bs4 if he wants it. The problem is really the *uninformed developer*: He needs our care and we should not give him a hard time if there's a simple way to avoid it.

**Comment by craigpotter** [comment]
Using Bs3 & bs4 in a project is an interesting issue.

One thing that crosses my mind though.... Why would you do that? 

It sounds like a massive headache trying to support 2 versions. 

BS3 isn't accepting new features. Just fixes.

**Comment by mikehaertl** [comment]
> One thing that crosses my mind though.... Why would you do that?

It's not that you *want* to do that. But large projects often develop their own dynamics. One reason I can think of: You have 2 customized Bs3 versions, one for frontend, one for backend with 1000s of customizations. Then one day management decides, that we need that one cool nifty feature from Bs4 in our frontent - tomorrow!!! If you now can update frontend to Bs4 independently of backend, you can lean back and relax. In the other case you're into serious trouble as you have to update your customized Bs3 backend, too.

**Comment by craigpotter** [comment]
If you have that level of customisation, it would either be in stylesheets or you won't be using the yii-bootstrap functionally surely ? If they have modified files in the Vendor folder, they are stupid. 
If they extended the classes, it shouldn't break but I mean its up to the core team how they want to handle it. 

If they move BS4 to a new repo, it has the possibility of having to replicate fixes on both repos. 

Be interested to here what @samdark  or @cebe think.

**Comment by mikehaertl** [comment]
> If they move BS4 to a new repo, it has the possibility of having to replicate fixes on both repos. 

That's not so much different from replicating fixes between branches. Also both versions may diverge, because incompatible Bs4 changes may be reflected in the yii2-bootstrap API at some point.

**Comment by jacmoe** [comment]
If BS4 is the current version of Bootstrap, then bs3 should just be a tag
and the project needs to move forward.
Alternatively, and probably the best way, a bs3 branch that will receive
only critical fixes should be made. And current version of BS for the main
branch would be BS4.

If people really have modified Bootstrap, that is their problem. They will
hopefully learn from their mistakes..
Den 25/05/2016 14.16 skrev "Michael Härtl" <notifications@github.com>:

**Comment by samdark** [comment]
Agree with @jacmoe and @mikehaertl.

**Comment by KAMAELUA** [comment]
Can someone explain to me, can I use bs4 now? If so, what I need to do? @samdark @jacmoe

**Comment by freezy-sk** [comment]
@KAMAELUA just use yii2-bootstrap dev-2.1 which uses Bootstrap 4.0.0.alpha2
but be careful because there are 991 commits since 4.0.0.alpha2 was tagged

**Comment by ffelaar** [comment]
Need a bit more assistance. I am including in composer.json the line:
`"yiisoft/yii2-bootstrap": "~2.1",`

I get an error that says :

> Problem 1
    - bower-asset/bootstrap dev-master requires bower-asset/jquery >=1.9.1,<=2 -> no matching package found.
    - yiisoft/yii2-bootstrap 2.1.x-dev requires bower-asset/bootstrap dev-default#48938155eb24b4ccdde09426066869504c6dab3c -> satisfiable by bower-asset/bootstrap[dev-master].
    - Installation request for yiisoft/yii2-bootstrap ~2.1 -> satisfiable by yiisoft/yii2-bootstrap[2.1.x-dev].

Please assist - i want to use bootstrap 4.

**Comment by webstylemedia** [comment]
Hey guys any news? In 2.1 still old bootstrap 3. :-(

**Comment by cebe** [comment]
should not be: https://github.com/yiisoft/yii2-bootstrap/blob/2.1/composer.json#L22

**Comment by webstylemedia** [comment]
cebe, I know about linking to that dev repository, but if you deploy it on your server and run test page and view source you will see that bootstrap.css and bootstrap.js has version 3.3.6.

I think this dev link in bower is wrong. It download 3.3.6

I used this command in separate project to get 4 alpha:
"twbs/bootstrap": "^4.0",

Please try it and see. Your current dev link is loading 3.3.6.

**Comment by bot4343** [comment]
I try install Yii2 with Bootstrap4-alpha2:
```
"require": {
        "php": ">=5.4.0",
        "yiisoft/yii2": ">=2.0.5",
        "yiisoft/yii2-bootstrap": "dev-2.1"
}
```
But got Composer error: `The requested package yiisoft/yii2-bootstrap dev-2.1 exists as yiisoft/yii2-bootstrap[2.0.0, 2.0.0-alpha, 2.0.0-beta, 2.0.0-rc, 2.0.1, 2.0.2, 2.0.3, 2.0.4, 2.0.5, 2.0.6, 2.1.x-dev, dev-master, 2.0.x-dev] but these are rejected by your constraint.`

What I'm doing wrong?

**Comment by rob006** [comment]
You should use:
```
"yiisoft/yii2-bootstrap": "2.1.x-dev"
```

**Comment by webstylemedia** [comment]
It still not work for me. I tried "yiisoft/yii2-bootstrap": "2.1.x-dev" too but still see Bootstrap v3.3.6 in source files....

It has yii2-bootstrap package files prepared for bootstrap 4 but from bower installed 3.3.6 version of bootstrap itself.

I attached screenshot.

Please help me.
![screenshot](https://cloud.githubusercontent.com/assets/650919/16782113/d4de5e00-4886-11e6-8037-0b6fcd3a7db6.jpg)

**Comment by bot4343** [comment]
I have the same, **2.1.x-dev** = **Bootstrap 3.3.6**

**Comment by tbreuss** [comment]
Same problem here. My settings are:
~~~php
    "minimum-stability": "dev",
    "require": {
        ...
        "yiisoft/yii2-bootstrap": "2.1.x-dev",
        ...
    },
~~~~
I updated composer and deleted the composer cache.
After `composer update` i have bootstrap 3.3.7.

**Comment by cebe** [comment]
I have just updated the version constraint and I am able to install bootstrap 4 with the following composer.json:

```json
{
    "require": {
		"yiisoft/yii2": "~2.0.9",
		"yiisoft/yii2-bootstrap": "~2.1.0@dev",
		"bower-asset/bootstrap": "~4.0.0@alpha",
		"bower-asset/jquery": "2.2.*"
    }
}
```

When not including the `bower-asset/bootstrap` requirement in my own composer.json however the installation does not work. Not sure what is going on with composer, which seems to apply the minimum stability in depdencies and does not allow overriding it with `@alpha`.

**Comment by cebe** [comment]
```
$ composer show
Warning: This development build of composer is over 60 days old. It is recommended to update it by running "/usr/local/bin/composer self-update" to get the latest version.
bower-asset/bootstrap        v4.0.0-alpha.3    The most popular front-end framework for developing responsive, mobile first projects on the web.
bower-asset/jquery           2.2.4             
bower-asset/jquery.inputmask 3.2.7             jquery.inputmask is a jquery plugin which create an input mask.
bower-asset/punycode         v1.3.2            
bower-asset/tether           v1.2.4            A client-side library to make absolutely positioned elements attach to elements in the page efficiently.
bower-asset/yii2-pjax        v2.0.6            
cebe/markdown                1.1.0             A super fast, highly extensible markdown parser for PHP
ezyang/htmlpurifier          v4.8.0            Standards compliant HTML filter written in PHP
yiisoft/yii2                 2.0.9             Yii PHP Framework Version 2
yiisoft/yii2-bootstrap       2.1.x-dev c048258 The Twitter Bootstrap extension for the Yii framework
yiisoft/yii2-composer        2.0.4             The composer plugin for Yii extension installer
```

**Comment by cebe** [comment]
okay, here is why the bootstrap package has to be included in the root composer.json when you have minimum stability set to stable: stability flags (`@alpha`) are root only and are ignored inside dependencies:

> require and require-dev additionally support stability flags (root-only).
> https://getcomposer.org/doc/04-schema.md#package-links

**Comment by cebe** [comment]
An other way to go would be:

> Alternatively, you can combine minimum-stability: dev with prefer-stable: true that will prevent everything from updating to dev versions but only take dev when necessary.
https://github.com/composer/composer/issues/4823#issuecomment-174461647

**Comment by set-st** [comment]
I'm trying move my project from mac to linux machine, and got this error:
```
$ composer update
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - The requested package bower-asset/bootstrap ~4.0.0@alpha exists as bower-asset/bootstrap[dev-master] but these are rejected by your constraint.
  Problem 2
    - yiisoft/yii2-bootstrap 2.1.x-dev requires bower-asset/bootstrap ~4.0.0@alpha -> no matching package found.
    - yiisoft/yii2-bootstrap 2.1.x-dev requires bower-asset/bootstrap ~4.0.0@alpha -> no matching package found.
    - Installation request for yiisoft/yii2-bootstrap ~2.1.0@dev -> satisfiable by yiisoft/yii2-bootstrap[2.1.x-dev].
```
On mac os composer update performs without errors. Any help, please.
```
    "minimum-stability": "stable",
    "require": {
        "php": ">=5.4.0",
        "yiisoft/yii2": "~2.0.9",
        "yiisoft/yii2-bootstrap": "~2.1.0@dev",
        "bower-asset/bootstrap": "~4.0.0@alpha",
        "bower-asset/jquery": "2.2.*"
    },
```

**Comment by samdark** [comment]
Try

```
    "minimum-stability": "dev",
```

**Comment by craigpotter** [comment]
With BS calling it on v3 (Well not completely but wanting to focus on v4) should we be looking at doing the same for yii2-bootstrap?

**Comment by cebe** [comment]
as v4 is not even released we will of course continue to support v3.

**Comment by skotos** [comment]
Just for the mix ...  yii\widgets\LinkPager is currently built around BS3 and is placed in the core Yii2 repo.  I've written a wrapper class that extends the LinkPager widget with BS4 stuff (maintains the A tag for disabled page buttons and overrides class name variables).  As bs4 takes more steam, should such a wrapper class be included in yiisoft/yii2-bootstrap(~2.1.0), or should the LinkPager in yiisoft/yii2 be modified to be more flexible with bootstrap versions?

**Comment by marnap89** [comment]
Hi :) Do you have any clue how to upgrade bootstrap to the newest alpha 5 version with composer?

**Comment by cebe** [comment]
does the [version constraint](https://github.com/yiisoft/yii2-bootstrap/blob/c048258180f6e8ac1149957585a2cddcf124d4c0/composer.json#L22) not match that version?

**Comment by pkbhadkoliya** [comment]
No it does not.

**Comment by tekord** [comment]
Try `composer require twbs/bootstrap:4.0.0-alpha.5` (it is official way from http://v4-alpha.getbootstrap.com/getting-started/download/)

**Comment by marnap89** [comment]
@tekord But it is not Yii way :)

**Comment by tekord** [comment]
@marnap89 it works for me very well. I set "@yii2/bootstrap" alias to my folder where I override classes to use BS4.

**Comment by marcocianci** [comment]
@tekord  and @marnap89 
to get latest bootstrap, run from root project folder: 
$ bower install bootstrap#v4.0.0-alpha.6

**Comment by skotos** [comment]
Is there a reason for holding tether at 1.2.x?  bs4-alpha6 in bower requires at least 1.4.x.  If we bump this in composer.json, alpha6 should just pull in. (at least I think that is whats going on.)

**Comment by cebe** [comment]
@skotos I do not think there is a reason, we are just not closely following bootstrap development to notice that. Could you test it with 1.4.x and send a PR if it works?

**Comment by Enrica-r** [comment]
I read the discussion about migration bs3 to bs4. I'm really a little bit worried. Boostrap has changed too many things completely so that all extensions and all applications couldn't be changed in one step. One example is Glyphicon. This has been remove in bs4. It's almost everywhere in Yii2 extensions. (BTW here is a temporary solution for this: https://gist.github.com/planetoftheweb/5d75a1ad45eb3059710747a3695fc068)

Probably same could happen with bs4 to bs5 in few years and so on.  If a change from one release to another is too hard the longer an old version will be kept in production. I'm afraid that too much of developers will keep 2.0.x caused by bs3. It's not good to have 3 LTS versions to maintain (Yii1.1, Yii2.0, Yii2.1...)

I wish (only wish not a request) that extension developers get time to change from now on and app developers could change page by page in their app. I think best process would be:

- Yii2 is be able to load bs3 (repo `bootstrap` as is) and bs4 (repo `bootstrap4`) in parallel.
- The developer can define in config which one is loaded in app's asset bundle by default. Without definition it's bs3 as is.
- The app developer should be able to set on each view which version will be loaded (eg. default bs3 if set bs4 in view --> app's asset bundle uses bs4 for this view).
- Extensions like @skotos LinkPager or Kartik's ActiveForm should be able to read from Yii app which bs versions are available and which one will be loaded in current view. The extension provider could build their wrappers to handle both versions.
- Mark bs3 as "depreciated" in Yii2 documentation for version 2.1.x and say in Yii's roadmap by which release bs3 isn't supported anymore.
- This should be done in Yii 2.0.x already because the extension developers need some time to update their code to make it compatible. With Yii 2.1.x (and productive bs4) application developers should start to update their pages.
- Remove bs3 in version >=2.2.0

I afraid that a hard change for all (extension developers and app developers) at same time will end up in a chaos or 2.0.x will be frozen until all important extensions have prepared a 2.1.x version.

Please think about the migration process. LTS versions of Yii should be caused by have Yii changes only and not by bs changes. Thanks to all.
Enrica

**Comment by Sammaye** [comment]
Glyphicons do get removed from time to time, originally in BS3 they didn't have them. Normally it is a fight between Bootstrap makers and the glyphicon makers for some legal issue or something

**Comment by dillix** [comment]
Its time to use Fontawesome instead of Glyphicons.

**Comment by Enrica-r** [comment]
@dillix I fully agree with you. Fontawesome should be used by default. Another posibility is Material Design Icons see https://materialdesignicons.com/.  Here you could compare the most important Icon providers: http://tagliala.github.io/vectoriconsroundup/

@Sammaye If the license question isn't fully clarified Glyphicons will be removed. 

Generally the icon thing was an example only.

**Comment by cebe** [comment]
> It's not good to have 3 LTS versions to maintain (Yii1.1, Yii2.0, Yii2.1...)

This will be only the extension to maintain, not the whole framework. You can use Yii 2.1 together with bootstrap extension 2.0 if you want to stay with bootstrap 3. Its not hard to maintain a bootstrap extension for bootstrap 3 if bootstrap 3 does not change.

Also I do not think there is a need to upgrade an application that is build on bootstrap 3 to bootstrap 4 as that would imply a rewrite of the CSS styles for the site.

Imo there is no need to think about an upgrade process for a CSS framework if you have a lot of custom code style. If you have mostly original bootstrap rules the upgrade should be easy.

**Comment by Enrica-r** [comment]
@cebe (Hello, I try to answer in english even if a discussion in German would be easier for me.)
I fully agree that the core of the framework isn't touched by this. Even the level BS repo is easy to handle with a parallel version 2.0 and 2.1 from your point of view. 

The hundreds of independent developed extensions are the critical point. They combine core, bootstrap. They uses boostrap classes for grids, inputs, buttons etc. actively. Yiisoft has one package coordinated with BS base and your BS extensions. Here is no issue. 

A single producer of extensions like Kartik could find his versioning strategy as you did with having two separate development lines.

But think about an app developer using extensions from different vendors. If each vendor has it's own strategy, versions and roadmap it will end up in a mess. So it's not an issue to change own app code from BS3 to BS4 but it's an issue to coordinate the extensions. But an app developer can't change before all his used extension have BS4 version ready.

For me as app developer Yii2 is ONE bundle. Yiisoft could help to do migration from BS3 to BS4 for all his dependent vendors. The core will be untouched anyway but Yii's bootstrap repo could help. If nothing is done on code level then Yiisoft should write a migration documentation with a migration instruction or recommendation for all affected vendors of extensions like:

- Build two repos or have two stable versions lines in Github and maintain both until a certain roadmap point (also valid for your yii2-bootstrap, keep 2.0 and 2.1 stable)
- Each affected vendor has to document the composer settings for BS3 rsp. BS4
- Declare BS3 as depreciated after a certain point (BS4 must be productive surely)
- Define a roadmap date or version on which all extensions must support BS4 only

Thank you very much for engagement.

**Comment by cebe** [comment]
> So it's not an issue to change own app code from BS3 to BS4 but it's an issue to coordinate the extensions.

we can write up some recommendation guidelines for this, so it is easier for extension developers to create something that is consistent.

> But an app developer can't change before all his used extension have BS4 version ready.

In most cases there are not many things to change from bootstrap 3 to 4 so if you rely on an extension you can work together with the author to make it compatible with bootstrap 4.

**Comment by Enrica-r** [comment]
@cebe Thanks for your support, I think you saw the issue. A recommendation is OK. It's not many to do but it's not compatible to each other.  Any way if all are aware it will work.

**Comment by samdark** [comment]
http://blog.getbootstrap.com/2017/08/10/bootstrap-4-beta/

**Comment by dryu** [comment]
Bootstrap 4 has progressed to Beta now.

**Comment by mitloshuk** [comment]
Bootstrap 4 autoupdated to Beta from composer, but:
composer.json
`"bower-asset/bootstrap": "~4.0.0@alpha",`
`"yiisoft/yii2-bootstrap": "2.1.x-dev",`

When using Bootstrap now
`Error: Bootstrap dropdown require Popper.js (https://popper.js.org)`

For example, Gii frontend not working correctly
`TypeError: $hint.parent(...).find(...).addClass(...).popover is not a function`

**Comment by dillix** [comment]
@mitloshuk, Bootstrap 4 has been migrated from tether.js to popper.js for tooltips, dropdowns & etc.

**Comment by samdark** [comment]
http://blog.getbootstrap.com/2017/10/19/bootstrap-4-beta-2/

**Comment by samdark** [comment]
It's in 2.1 branch for a long time. Will be tagged when v4 will be released.

**Comment by samdark** [comment]
http://blog.getbootstrap.com/2017/12/28/bootstrap-4-beta-3/

**Comment by bobonov** [comment]
The release of yii2 bootstrap4 should be done together or after issue yiisoft/yii2#10712 since bootstrap4 has dropped glyphicon

**Comment by berosoboy** [comment]
Bootstrap 4 was released today

**Comment by samdark** [comment]
http://blog.getbootstrap.com/2018/01/18/bootstrap-4/

**Comment by esaesa** [comment]
18 Jan 2018, Bootstrap Stable released.

**Comment by dillix** [comment]
@bobonov May be we should use fontawesome as drop in replacement for glypicons?

**Comment by acerix** [comment]
For the glyphicon issue, I suggest using an approach like this which maps them to fontawesome:

https://gist.github.com/blowsie/15f8fe303383e361958bd53ecb7294f9

Here's a demo of that with a basic grid view:

https://codepen.io/acerix/full/Eorazw/

**Comment by dillix** [comment]
@acerix there is no glyphicons in bootstrap 4 anymore, it was dropped for license issues.

**Comment by Insolita** [comment]
So, as i see yii2-bootstrap with bootstrap4 coming in 2.1 branch... and this make impossible to use both 3.x and 4.x version in one project (3.x with adminLte for backend and 4.x for frontend) without duct tape

**Comment by samdark** [comment]
That's valid case...

**Comment by rob006** [comment]
This was discussed almost 2 years ago. Using separate extension for Bootstrap 4 may simplify using two different versions of Bootstrap in single project, but it will make upgrade from bootstrap3 to bootstrap4 more difficult (you need to change namespace everywhere) and maintain extensions that depend on bootstrap will become a nightmare (you need to support 2 lines even if the only difference is used `yii2-bootstrap` extension). There is no perfect choice here.

**Comment by Insolita** [comment]
Hmm, the decision about change bootstrap version in existed project is more complex, is not just about change yii2-bootstrap version, it can depends on many third-party plugins and requires significant changes on the part of the design layout. The replacing namespaces is not so difficult task in modern IDE. 
As matter of  fact bootstrap-3 based solutions there are very popular for backend-side, and it is enough for most  backend cases. But front-side often need to use more modern trends
But is only my opinion.

**Comment by schmunk42** [comment]
> maintain extensions that depend on bootstrap will become a nightmare (you need to support 2 lines even if the only difference is used yii2-bootstrap extension).

Using extensions will become a nightmare if it is in the same namespace, if one is requiring `2.0.*` and another `2.1.*` it'll be impossible to update.

The use-cases mentioned by @Insolita are very common and important. In contradiction to "normal" PHP libs, this one *could* exists multiple times, since it's (almost) only a wrapper for an asset package.

**Comment by Enrica-r** [comment]
I have anticipated the upcoming issue what you discussed few posts again. Look back to my post of 16th April 2017 already. At this date the time wasn't mature for such a discussion because all developped with BS3 and was simple to fork a separate GIT version.

I support the concern that not all extensions using BS3 (like Kartik's) at the same shot. Also it's almost impossible to change a larger application in one release. Yii2 risks that 2.0.x release will be used for long term.

So I proposed to install a B3 extension (marked as depreciated) parallel to BS4 extension. In release 2.1 Yii2 should ensure that "old" extensions work as usual with BS3. After with release 2.1 all extension developers should get time to migrate their extensions to BS4. If they want they could support parallel installation also. Yii2 should deliver a function which lists the installed BS versions and which BS version is default. With release 2.1.x BS4 should be default and with 2.x.0 BS3 will be removed.

I know that such a migration/versioning concept will produce a lot of work but it can use for future BS or Jquery versions.

@samdark : I don't what the correct way is. I don't want criticize Yii2 developer as it seems sometimes. I thank for your engagement. Please check my proposal of 16/04/2018 to avoid the "nightmare".

**Comment by rob006** [comment]
@Insolita I've migrated yii1 app from bootstrap2 to bootstrap3: 
* for backend it was relatively simple (few hours of work), since most of work was done by php-wrappers,
* for frontend impact of used php-wrapper was irrelevant due to many bootstrap customizations.

If I will need to use separate package with different namespace it will be much harder for me.

@schmunk42 
> Using extensions will become a nightmare if it is in the same namespace, if one is requiring 2.0.* and another 2.1.* it'll be impossible to update.

Releasing separate package for bootstrap4 will not save you this nightmare. There are over 1k extensions that depend on this package - all of them should have a clone for bootstrap4, so you could use the same extension for two different versions of bootstrap in a single app. It's unrealistic, a separate package will only increase fragmentation (it will be harder to maintain extension for two different versions of bootstrap).

**Comment by schmunk42** [comment]
> all of them should have a clone for bootstrap4, so you could use the same extension for two different versions of bootstrap in a single app.

Sorry to say that, but that's wishful thinking - it *should* be that way, but it's not going to happen very fast and smoothly. By using the same repo we'd just pretend it would be easy, but it's making things much worse.

---

Here's a list of our current dependencies to `yii2-bootstrap`:

```
root@714a18efd4d3:/app# composer depends yiisoft/yii2-bootstrap
2amigos/yii2-selectize-widget     1.1.0   requires  yiisoft/yii2-bootstrap (~2.0.0)  
dektrium/yii2-user                0.9.13  requires  yiisoft/yii2-bootstrap (^2.0.0)  
dmstr/yii2-adminlte-asset         2.6.0   requires  yiisoft/yii2-bootstrap (~2.0.0)  
dmstr/yii2-bootstrap              0.2.1   requires  yiisoft/yii2-bootstrap (^2.0.6)  
dmstr/yii2-cookie-button          0.2.0   requires  yiisoft/yii2-bootstrap (*)       
kartik-v/yii2-krajee-base         v1.8.9  requires  yiisoft/yii2-bootstrap (@dev)    
lajax/yii2-translate-manager      1.7.1   requires  yiisoft/yii2-bootstrap (*)       
dmstr/phd5-app                    -       requires  yiisoft/yii2-bootstrap (2.*)     
yiisoft/yii2-apidoc               2.1.0   requires  yiisoft/yii2-bootstrap (~2.0.0)  
yiisoft/yii2-debug                2.0.13  requires  yiisoft/yii2-bootstrap (~2.0.0)  
yiisoft/yii2-gii                  2.0.6   requires  yiisoft/yii2-bootstrap (~2.0.0)  
zhuravljov/yii2-datetime-widgets  1.1.2   requires  yiisoft/yii2-bootstrap (~2.0.0)  
```

Just take `yii2-apidoc` as an example, I might just want to create docs with BS3, but the rest of my app will use BS4 (even if I'd have to fork 5 repos for that). How would I do this? I can not install two versions of the same extension, even if this is a `dev-dependency`.

---

Instead, as mentioned before, when upgrading from BS3 to BS4 in an extension or app, I'd change

    use yii\bootstrap\Collapse;

into

    use yii\bootstrap4\Collapse;

**Comment by bobonov** [comment]
Something to consider in having 2 separate repos.
Scenario:
You start using porting BS4, you correct all your namespace to point to BS4.
You don't correct vendors namespace and fix the extensions to use BS4 classes. 
Therefore if you use 3rd party extension that still use bootstrap this will lead to a big mess since BS3 and BS4 are loaded at same time overriding each other in term of css rule and probably causing js conflict to each other.

So porting to BS4 is not just matter of changing namespace, it is a app wide work.
You can operate per single view, having one view all BS4 while another is still BS3 but think at the result, you will have 2 pages that look differently, maybe slightly but different. So if you port to BS4 you can't go in production till you finish

While some yii2 extension should work out of the box just by using newer version of yii2-bootstrap some other depends on a third party extension to be ported first (ie select2)

But the issue about front/back is true, single repository do not allow to work on them separately.

In this case I would reconfigure yii2 to use 2 separate composer.json/vendors for front and back.
Something like
frontend/vendors
frontend/composer.json
I know this is not ideal and nice but this would allow to keep separated the old AdminLTE BS3 backend from the new shining BS4 frontend without accidents.

**Comment by rob006** [comment]
> By using the same repo we'd just pretend it would be easy, but it's making things much worse.

By using separate repo we'd just pretend that you could simply use bs3 i bs4 simultaneously, but this is far more complicated. 

> How would I do this?

You don't need to add `yii2-apidoc` to dependencies in the first place, you could install it separately. ;)

Anyway, both solutions are bad. Instead arguing which is worse we could try to create one extension compatible with both versions of bootstrap with switch like `Bootstrap::setVersion(Bootstrap::V4))` which changes output generated by helpers/widgets.

**Comment by schmunk42** [comment]
> By using separate repo we'd just pretend that you could simply use bs3 i bs4 simultaneously, but this is far more complicated.

The use-case is not to use them on the same page/layout, but having a backend modules in BS3 and a frontend in BS4 will be a very common thing. If this is not possible, It's like restricting the application just to use one "theme".

> Instead arguing which is worse we could try to create one extension compatible with both versions of bootstrap with switch like Bootstrap::setVersion(Bootstrap::V4)) which changes output generated by helpers/widgets.

It won't result in cleaner code, if one extension could support both versions simultaneously.

**Comment by Enrica-r** [comment]
In my opinion the Yii team has to avoid a hard break analogue Yii1 to Yii2 caused by BS version change. Unfortunately Yii depends on BS in most extension. I aggree with @schmunk42 to have a clean code but the Yii team should ensure a coordinated migration from BS3 to BS4 (incl. JS).

As written in other posts BS3 and BS4 can't be mixed on one page (layout) but it could be mixed in an application with two layouts (eg. frontend vs backend).

In my opinion there should be clear migration phases defined by Yii and this must be obligatory for all extension developers otherwise an extension will loose its Yii2 2.1.x compatibility. My proposal is to have four phases: (@samdark )

1. Yii2 prepares a release which supports two repos as mentioned before (another namespace) plus a method in 'Yii::$app' which delivers BS version loaded in current layout's asset bundle. Yii team publishes migration instructions with examples to extension developers and app developers. A published roadmap for developers is obligatoy also.

2.  Extension developers get 3 to 6 months time to adapt their BS dependent extensions. They have to document the compatibility and how it works. They could have to alternatives to do this:
a) Maintain two forks distinguished by namespace. Adv: clear code, clear decomm / Disadv: maintain to branches parallel for a certain time.
b)  Support both BS version in one code. The extension asks 'Yii::$app->bsVerison' and reacts concerning the version. Adv: one namespace, no parallel packages to maintain, easier for app developers / Disadv: Makes code more complicated, needs a cleanup release later for each extension.

3. After 6 months (hopefully all extensions will have been adapted) app developers adapt their applications. If they do all work in one relase or if they migrate FE/BE separately or if they change some areas is their decision. This phase could have a period of 1 - 2 years.

4. After phase 3 Yii team announces a definitive decommision of BS3 support. The package 'yii-bootstrap" will be removed and also extension developers are allowed to remove their packages or cleanup theirs codes from BS3 part.

Probably Boostrap will develope version 5 already in this time and the process will start again. Same procedures could be used for Jquery upgrades. @samdark : Please discuss with your core team, how Yii could support BS version change (also for their own extension).

**Comment by bobonov** [comment]
> As written in other posts BS3 and BS4 can't be mixed on one page (layout) but it could be mixed in an application with two layouts (eg. frontend vs backend).
I think most of the people misunderstood my issue about mixing BS3 and BS4 in same view, bad on me and on my English.

Some extension will work out of the box since they will inherit their bootstrap code directly from yii2-bootstrap*.
Some other uses tags and classes directly and need to be migrated first.
Some other extension relay on third party code to be migrated first, ie any yii2-select2* widget

I think that we all agree that (beside the yii2-boostrap solution will be) my migration can't finish till all the extension I use get migrated.

yii2-bootstrap new version VS yii2-bootstrap and yii2-bootstrap4
Many point are have already been exposed in the discussion, my concern about yii2-bootstrap and yii2-bootstrap4 is that would be more difficult to do a work in progress migration.

What I mean.
I use Select2 in my project. Probably I use it in many forms and in many search forms.
I start migration so I replace yii2-bootstrap with yii2-bootstrap4 in composer.
But yii2-bootstrap is not removed since my third party yii2-selec2 widget require it.
So I obtain that BS3 and BS4 get mixed and every page that use the select2 widget stop loading properly and probably is not usable at all.
On the other way with yii2-bootstrap version bumping maybe my Select2 looks weird or in the worst case is not working but my page should load and I can work  on porting the rest of the elements while waiting for a new Select2 release which support bootstrap.
In this way the work can go in parallel with yii2 widget fixing and I do not have to wait.
Cons it that you need to use separate composer (in the worst case one for each application you have)

**Comment by schmunk42** [comment]
> But yii2-bootstrap is not removed since my third party yii2-selec2 widget require it.
So I obtain that BS3 and BS4 get mixed and every page that use the select2 widget stop loading properly and probably is not usable at all.

That's true and it's an issue. `AssetBundle::conflicts` comes to my mind to handle that, but it's just an idea.

What's a bigger issue is, that if you want to install ie. a newer version of `gii` which depends on BS4, you won't be able to do that if just one single input in your application requires BS3.

**Comment by schmunk42** [comment]
I need to reference https://github.com/yiisoft/yii2-bootstrap/pull/212 here, it's about **removing bower dependencies completely** and not managing assets like JS/CSS with `composer`.

IMHO this should be a goal for the next release, since the framework core in `2.1` won't include bower-packages anymore. Removing those dependencies in `composer.json` might also help those who are planning to install different BS versions for different layouts, since there won't be any conflicts - now it's the time for some bigger changes 😄

**Comment by JQL** [comment]
Opening a can of worms here but, why should "Yii" be dependant on any CSS framework?

Wouldn't it be better in the long run to move the CSS frameworks into libraries and the user decides which framework to use? With a system default if none is chosen.

OK, I know there's more than a ton of work into setting up the libraries and the code to call the correct libraries.

My 2¢

**Comment by jacmoe** [comment]
It's not dependent on any CSS framework :)

That said, yes: some features of Yii  - debug interface - depends on Bootstrap.
To make life easier, the Yii team decided on Bootstrap instead of rolling their own.

I tend to not use any Bootstrap related Yii functionality, though.

**Comment by JQL** [comment]
When you look at the code for the Widgets, Tables etc. they include BootStrap3 mark up:

> /**
>      * @var array the HTML attributes for the container tag of this widget. The `tag` option specifies
>      * what container tag should be used. It defaults to `table` if not set.
>      * @see \yii\helpers\Html::renderTagAttributes() for details on how attributes are being rendered.
>      */
>     public $options = ['class' => 'table table-striped table-bordered detail-view'];

Why not have a system to translate _table_ into say "_display:grid_" based of whether the user wants CSSGrid or BootstrapX or some other framework?

**Comment by jacmoe** [comment]
If you don't set the options, default values are assigned.

2018-02-10 21:55 GMT+01:00 John Lavelle <notifications@github.com>:

**Comment by schmunk42** [comment]
I talked to @cebe a while ago about this.

My suggestion would be to create a new repo ie. `yii2-bootstrap4` which is only compatible with 2.1; including new asset management and Bootstrap 4.

You can still decide if this repo should support BS4 in a later version, but now it looks rather tied to BS3 (see https://github.com/yiisoft/yii2-bootstrap/issues/221#issuecomment-365178860)

**Comment by FabrizioCaldarelli** [comment]
I'm agree with @schmunk42 because developers should be able to decide if upgrade to bootstrap 4 or keep version 3.
I have too much projects based on bootstrap 3, so if I had to upgrade every single project this would be a lot of work.

**Comment by FabrizioCaldarelli** [comment]
The matter to making a soft transition from bootstrap 3 and bootstrap 4 is very important.

This could be a possible approach.

Bootstrap 3 should not be updated frequently anymore, so we could package the bs3 sources statically inside in a yii2-bootstrap3 package,
that we should update everytime there is an update of version 3 (I think not frequently).

This is needed because bootstrap composer package "twbs/bootstrap" now points to version 4.

Then we have to create a yii2-bootstrap4 package that contains a dependency to "twbs/bootstrap", so it will install last bootstrap version 4.

The yii2-bootstrap3 package will expose a Boostrap3Asset, and the package yii2-bootstrap4 will expose a Bootstrap4Asset.

When twitter will release bs5, we should only put statics bs4 sources inside yii2-bootstrap4 package and create a new yii2-bootstrap5 that will point to same repository "twbs/bootstrap" (if it will not change), and so on.

The developers will can decide what bs version to use, without creating panic.

**Comment by FabrizioCaldarelli** [comment]
To demonstrate previous post, I've create 2 extensions:

- https://github.com/FabrizioCaldarelli/yii2-bootstrap3 that contains bs3 source code inside and provides Bootstrap3Asset.php;
- https://github.com/FabrizioCaldarelli/yii2-bootstrap4 that requires twbs/bootstrap package and provides Bootstrap4Asset.php;

Then I removed references to Bootstrap3 to yii2-bootstrap extension: https://github.com/FabrizioCaldarelli/yii2-bootstrap/commit/3de2226520f9198feed11b8c2a1fda313bed4009 
This is needed to avoid to avoid breaking change and keep compatibility with existing widgets/extension.

Similar at this, we can change in yii2-gii references to bootstrap3 specifying, for example:
- 'sfmobile\ext\bootstrap4' or 'sfmobile\ext\bootstrap3' instead 'yii\bootstrap\BootstrapAsset' at https://github.com/yiisoft/yii2-gii/blob/master/src/GiiAsset.php#L29


Finally I've create a fork from yii2-basic-app with these changes:

- Added two layouts : main-bs3.php, main-bs4.php, that apply respectively Bootstrap3Asset and Bootstrap4Asset;
- Added a site/test-bs4 action to show how same controller can use different bootstrap versions;

https://github.com/FabrizioCaldarelli/yii2-app-basic-bs/commit/3b7f7fb37d6c66cc407f0d902d1bb5699ae9f930

**Comment by Enrica-r** [comment]
@FabrizioCaldarelli 
Thank you for your good work an your concept. Your idea with two Layouts 'main-bs3.php' amd 'main-bs4.php' you can decide for each page which boostrap version will be loaded. So far ok.

In a further step depending widgets (extension) should be adapted also. As a developer of such widgets we have two possibilities:
- first: The view (layout) or Yii's app has a property which tells my wigets which version of BS (or better CSS framework) is loaded so that the widget can put out different code Syntax or it produces an error like 'this Framework isn't suported'
- second: For each widget package exists a bs3 version as a b4 version also (similar than your packages).

Using the version tags like 'Yii-bootstrap' does ends up in a problem. Version 2.0.x suports bs3 only, version 2.1.x suports bs4 only but version 2.2.x bs3 only again. This is confusing and the maintenance effort is unacceptable. With proposal one 'Yii-boostrap' could provide in version 2.2.x b3 and b4 compatibility.

I prefer alternative 1 but then you package should deliver a framework Name like 'bootstrap-3' accessible for widgets.

Your work should be part of Yii's bootstrap package. Coordinated for all developers.

**Comment by FabrizioCaldarelli** [comment]
@Enrica-r  From widgets point of view, I also agree with your first proposal.

We could add a "uiFramework" property to View class, so the layout used could specify what  ui-framework has used. This could be useful even if we want to use other ui framework (semantic ui, etc..) and the widget would adapt to it.

What do you think?

**Comment by Enrica-r** [comment]
@FabrizioCaldarelli : That's exactly my alternative no. 1 and also what I wrote on 22 and 23 Jan. this year.

This property shouldn't be in your package. It should be in Yii's core. If you look back to Yii2's history this wasn't an issue because BS3 was almost part of Yii fully implemented. Slowly the strategy changed and Yii core moved all ui elements in separate packages so that Yii core is independent of any css framework. Theoretically it should work with ZURB foundation. 

If your ui property is part of yii's core all widget developers can check which css framework is loaded with the layout. So they can react with an error message or they can deliver translation class. (proposal of @JQL ).

@samdark Could you analyze @FabrizioCaldarelli 's solution? If it's a way then merge it to version 2.1.x and merge functionality of 'Yii-bootstrap' version 2.1 and 2.2 while using view's 'uiFramework' property. This seems to be the only way to avoid a 2.1.x and 2.2.x fork for longer duration.

**Comment by FabrizioCaldarelli** [comment]
@samdark  @Enrica-r Also this solution will allow to use different UI Framework. It is only requested to widgets developer to handle these framework, with little differences inside the same widget source code.

**Comment by Enrica-r** [comment]
@FabrizioCaldarelli Exactly what I think and what I discussed during BS4 alpha phase, but nobody was aware of it. Now it comes up.

**Comment by schmunk42** [comment]
> We could add a "uiFramework" property to View class, so the layout used could specify what ui-framework has used. This could be useful even if we want to use other ui framework (semantic ui, etc..) and the widget would adapt to it.

I don't think this is a viable solution, I will just lead to a mess in widget code.

**Comment by FabrizioCaldarelli** [comment]
@schmunk Alternatives to allow widgets developers to handle different ui framework without creating many package of same source one (one for ui framework)?

**Comment by schmunk42** [comment]
> Alternatives to allow widgets developers to handle different ui framework without creating many package of same source one (one for ui framework)?

None. PHP packages (or asset packages via composer) can only be installed in one version, but technically there's no reason for not installing `twbs/bootstrap` in 3.x and 4.x.

**Comment by FabrizioCaldarelli** [comment]
@schmunk I think that you, as all of us, have a lot of projects based on Yii 2.0 and bs3. How do you plan to switch to Yii 2.1 without updating  all views of all projects?

**Comment by schmunk42** [comment]
In our case I'd plan to continue using `yii2-bootstrap` (BS3) for the backend (AdminLTE).
And BS4 in the frontend via `yii2-bootstrap4` (if there'll be a package for it :))

I actually don't have the illusion that this can be done automatically.

**Comment by FabrizioCaldarelli** [comment]
@schmunk bs4 for new frontend projects. What about existings? What about using other ui framework (or eventually no ui framework)? All widgets are heavily linked to bs4 and yii does not offer convenient way to manage switching from an ui fw to another.

**Comment by schmunk42** [comment]
> bs4 for new frontend projects. What about existings?

I think switching involves some amont of manual work. For BS3 to BS4 I'd change the widget namespace from `yii\bootstrap` to `yii\bootstrap4`.

> What about using other ui framework (or eventually no ui framework)? All widgets are heavily linked to bs4 and yii does not offer convenient way to manage switching from an ui fw to another.

You gave the answer already :) Widgets are heavily linked to the CSS (and sometimes JS) framework. There's no easy way to switch between them, just by setting a property.

For example, I just picked this one randomly https://github.com/nonzod/yii2-foundation - some widgets are the same, some are only available for Zurb some others only for Bootstrap.

---

While it would be great if we could switch CSS framework so easily, I simply think it creates more work than doing a clean switch. I once [tried this](https://github.com/schmunk42/lessii) when we switched from blueprint to bootstrap six years ago, it was just not feasible.

**Comment by Enrica-r** [comment]
@schmunk42 You are worry about a mess within widget code. I say yes and no because it depends how complex a widget is. For simple widgets which uses some BS CSS classes it's easier to build in a "library" class which delivers BS3 or BS4 classnames depending on which framework is loaded in main layout (recognizable via a property in the view class). You save time to maintain several packages.

For more complex widgets using different Jquery versions it could be easier to handle separate repositories which are addressed by name space. The widget init method should verify framework version and throw an error if BS3 is loaded if BS4 is expected.

Fabrizios method allows to have eg. backend on BS3 and frontend on BS4 or it allows to do stepwise migration within a project. But it's not possible before extension developers haven't adapted their widgets (eg. Yii-Bootstrap or Kartik's active form).

**Comment by schmunk42** [comment]
> allows to do stepwise migration within a project.

My point is, that you shouldn't do a "stepwise" migration of a CSS framework, because it will mess things up - just use one CSS framework per layout. And this is also possible when having two extensions, but in a much cleaner way.

**Comment by Enrica-r** [comment]
@schmunk42 That's clear and I fully agree with you. One layout (eg. main-b3.php) can have only one css / js framework. But another newer part of your project can use (eg. main-b4.php). If the widget code can recognize this you don't have to do anything (concerning this widget). Otherwise change name space pointing to an adapted version.

**Comment by Enrica-r** [comment]
@schmunk42 After a parallel migration phase (defined by Yii project) all extension developers could remove old BS3 packages or code libraries. If Yii is producing a hard break BS3->BS4 old versions will survive too long.

**Comment by FabrizioCaldarelli** [comment]
@schmunk This is not only a stepwise matter, but there are many widgets (Kartik select, date picker, etc..) that they are not immediately available on bs4. And when Kartik will update them, it could point directly to Yii-bootstrap4, messing up where we use bs3. I think that we should give a guideline to handle these switches, because they could happen frequently.

**Comment by schmunk42** [comment]
> all extension developers could remove old BS3 packages or code libraries

Why should someone **remove** BS3 code? There's absolutely nothing wrong with it. Moreover, there's no central authority which can force any developer to do something.

>  And when Kartik will update them, it could point directly to Yii-bootstrap4, messing up where we use bs3. I think that we should give a guideline to handle these switches, because they could happen frequently.

The guidance is: Also **create a new extension** or **do a major release** - anything else is a bug. Btw: the former one would allow you to use those extensions with BS3 and/or BS4, while a major version upgrade would force you to use a specifc BS version.

**Comment by Enrica-r** [comment]
@schmunk42 
>Why should someone remove BS3 code? There's absolutely nothing wrong with it. Moreover, there's no central authority which can force any developer to do something.

I don't agree with this. All software provider terminate their support for certain als versions. If I were Kartik I would support BS3 packages or compatibility for two or three years. Another developer migrates his project and doesn't provide both mainstream versions parallel. What a mess!

That's why I proposed that Yii team defines a migration phase during wich all providers of extensions have to maintain BS3 and BS4. If you want to use BS3 for longer period after the "official" migration phase you will have to fork the extension.

**Comment by schmunk42** [comment]
> I don't agree with this. All software provider terminate their support for certain als versions.

For me removing code and dropping support are two different things.

> If I were Kartik I would support BS3 packages or compatibility for two or three years. Another developer migrates his project and doesn't provide both mainstream versions parallel. What a mess!

Thinking about this, it's one more reason to create a separate extension. We should not even start having properties in the views, triggering "magic" changes of widgets markup. Followed by the requirement that all widgets would have to be updated to support such a change.

> all providers of extensions have to maintain BS3 and BS4.

I don't want to annoy people by repeating myself, but this is simply not possible.

---

@samdark My position on this became more clear after the recent discussions: 

Close this issue as **wontfix** and create a new repo `yii2-bootstrap4` - and add a migration guideline to 2.1 that extension developers should do the same.

**Comment by samdark** [comment]
Yes. Makes sense.

**Comment by FabrizioCaldarelli** [comment]
@schmunk will yii2-bootstrap continue to pointing to twbs/bootstrap repository (locked to major version 3) or will it contain bootstrap source code inside?
Because if both yii2-bootstrap and yii2-bootstrap4 will point to bootstrap repository, there will be a conflict with file names, because bootstrap has maintained same vendor /twbs/bootstrap path.

**Comment by schmunk42** [comment]
> will yii2-bootstrap continue to pointing to twbs/bootstrap repository (locked to major version 3) 

I'd leave it pointing to bower/boostrap (v3).

`yii2-bootstrap4` will load its sources from npm. Depending on the way of the installation you can have multiple versions (foxy+npm) or just one (asset-packagist).

**Comment by kartik-v** [comment]
Was led to this issue from another one. Feel it makes sense to have a separate Yii2 extension for BS3 as well as BS4 - since BS4 includes some significant changes:

1.  especially in HTML markup for some critical components and way widgets will be rendered
2. includes changes to JS/JQuery dependencies
3. includes changes to icon framework dependencies (most default icons will change)
4. as well as changes to basic default CSS look and feel (e.g. flex layouts, font family, font size, line heights, obsoleting of basic CSS classes like `pull-right`, `btn-xs` etc.)

In this scenario all BOOTSTRAP based Yii2 extensions need appropriate dependencies to run both for BS3 and BS4 based on the situation and choice of the developer.

**Comment by samdark** [comment]
Considering all the feedback, I think it makes sense to have yii2-bootstrap4 separately.

**Comment by alexantr** [comment]
Yes! BS3 & BS4 like Yii 1.1 & Yii 2 - too different

**Comment by samdark** [comment]
Need to move 2.1 branch there.

**Comment by schmunk42** [comment]
btw: regarding https://github.com/yiisoft/yii2/issues/15983

It could (should) be released as **yiisoft/yii2-bootstrap4** version `1.0.0`

**Comment by kartik-v** [comment]
> Need to move 2.1 branch there.

👍  Great... maybe the namespace in composer.json needs to be changed to `bootstrap4` as well... so that it can be referred something like `yii\bootstrap4\Modal:;widget` for example

**Comment by samdark** [comment]
We're going to finally agree on global versioning policy soon. Not closing the issue till then.

**Comment by cchandel** [comment]
Can't find where this has been covered so ...
What's the correct way to move to bootstraps4 from a fresh install of advanced template?
I tried replacing yiisoft/yii2-bootstrap with yiisoft/bootstrap4: ~1.0.0 but I get errors that force me to change stability to dev.
Yet, I still see only bootstrap.css v3 being included in the pages being served.
What am I doing wrong?

**Comment by machour** [comment]
Can be closed

**Comment by samdark** [comment]
Not needed. 2.1 was never released.
